### PR TITLE
Preserve the contextvars context when offloading execution to a thread

### LINKED ISSUE #143: RFE: preserve contextvars when performing the request
The OpenCensus tracing framework has an extension (opencensus.ext.requests) that monkey-patches `requests.Session.request()` and logs all outgoing requests and their status to a metric collection system of your choice.  It relies on [contextvars](https://docs.python.org/3/library/contextvars.html) to keep track of the tracer that is used by the current thread/async context, so the outgoing requests can be attributed to the right trace.

By offloading the actual request execution to a worker thread, `requests_futures` breaks this integration, and so outgoing requests remain untraced.  This can be easily fixed by following [the suggestion in the contextvars PEP](https://peps.python.org/pep-0567/#offloading-execution-to-other-threads) and wrapping the calls with `functools.partial(contextvars.copy_context().run, func)`).

### PR DESCRIPTION
**Pull request by mgedmin**
This follows the advice of the contextvars PEP:
https://peps.python.org/pep-0567/#offloading-execution-to-other-threads

Closes #143.

I have tested this code in our application, and it fixes opencensus.ext.requests tracing of outgoing requests.

### CODE REVIEW

File: requests_futures/sessions.py [hunk 1]
```diff
@@ -115,6 +120,11 @@ def request(self, *args, **kwargs):
                 dumps(func)
             except (TypeError, PickleError):
                 raise RuntimeError(PICKLE_ERROR)
+        elif contextvars:
+            # contextvars.Context objects are not pickleable, so we cannot do
+            # this when using ProcessPoolExecutor
+            ctx = contextvars.copy_context()
+            func = partial(ctx.run, func)
```
Comment (ross) [inline requests_futures/sessions.py:127]:
```suggestion

        if self.context_callback:
            func = self.context_callbak(func, self, *args, **kwargs)
```

**Review by ross — COMMENTED** [review]
This looks simple enough and I'm not opposed to doing something to support the use case, but I'm a little hesitant to have something that's one-off and specific to `contextvars` since other people may need to do something else to address similar situations for other libs.

What would you think about adding a param to `FuturesSession.__init__`, maybe something like `context_callback`.  That should let you do something like:

```python
def copy_context(func, *args, **kwargs):
    ctx = contextvars.copy_context()
    return partial(ctx.run, func)

fs = FutureSession(..., context_callback=copy_context, ...)
```

**Comment by mgedmin** [comment]
Well, contextvars is a standard library facility meant for interoperability between all the async frameworks in Python, so I thought that supporting it by default seemed reasonable.

(There are at least two factors that detract from this argument: some non-async frameworks still use thread locals for which there's no copy_context() alternative, and lack of support for multiprocessing due to pickling issues, both make this a less universal solution than I would've liked.)

I will give a little thought to context_callback.  I wonder if it's even necessary, since I can provide my own executor wrapper that does the context_copy() thing in it submit() before forwarding to the real ThreadPoolExecutor.  I'll come back after doing some experiments.

**Comment by mgedmin** [comment]
I thought up a new argument against my PR (why am I like this?): some applications/libraries might not be expecting mutable objects referenced by their contextvars to be shared by multiple threads.

**Comment by mgedmin** [comment]
Here's a solution that doesn't require requests_futures code modification:

```python

import contextvars
from concurrent.futures import ThreadPoolExecutor
from requests.adapters import DEFAULT_POOLSIZE
from requests_futures.sessions import FuturesSession


class ContextPreservingThreadPoolExecutor(ThreadPoolExecutor):
    # opencensus uses contextvars to keep track of the current tracing span; we want to preserve
    # that context so opencensus.ext.requests will know which incoming request caused an outgoing
    # request to a dependency.  If we don't do this, opencensus.ext.requests cannot find the
    # right tracer in the contextvars, so it gets a NoopTracer and discards any traces without
    # submitting them to App Insights.

    def submit(self, fn, /, *args, **kwargs):
        # https://peps.python.org/pep-0567/#offloading-execution-to-other-threads
        ctx = contextvars.copy_context()
        return super().submit(ctx.run, fn, *args, **kwargs)

...

max_workers = 8  # or whatever you want
session = FuturesSession(
    executor=ContextPreservingThreadPoolExecutor(max_workers=max_workers),
    adapter_kwargs={
        'pool_connections': max_workers,
        'pool_maxsize': max_workers,
    } if max_workers > DEFAULT_POOLSIZE else {},
)
# now do things with session
```

This could maybe be converted into a documentation patch?

One caveat: this leaves session._owned_session = False, so using it as a context manager will not shut down the thread pool and cancel outstanding requests.  I'm not sure how to solve that nicely.

**Comment by ross** [comment]
> There are at least two factors that detract from this argument: some non-async frameworks still use thread locals for which there's no copy_context() alternative,

This was the main point behind my thinking, in combination with the potential for someone to need to do something other than muck with context for some reason.

> Here's a solution that doesn't require requests_futures code modification:

That seems clean and smooth. 

> This could maybe be converted into a documentation patch?

Not opposed to it being documented since you ran across the need for it and it wasn't immediately clear the best way to go about it. 

> One caveat: this leaves session._owned_session = False, so using it as a context manager will not shut down the thread pool and cancel outstanding requests. I'm not sure how to solve that nicely.

Yeah, that's the main downside of passing in an executor, `_owned_executor` will be False and it'll be on the caller to shut it down if/when they close the session.

**Pull request closed without merging**